### PR TITLE
CORTX-29888: Upgrade mini-provisioner interface (config) for rolling …

### PR DESCRIPTION
**Pull request by atulsdeshmukh2312**
…upgrade

Signed-off-by: Atul Deshmukh <atul.deshmukh@seagate.com>

# Problem Statement
-  Upgrade mini-provisioner interface (config) for rolling

# Design
Config upgrade:

Each component provides a sample config file which is installed/copied as the main config for the first time. However when the CORTX component is upgraded, the main config exists and can not be overwritten. There are 3 possible scenarios that needs to be handled.

Add Keys - For this case, only new entries should be added to main config along with the default value.
Replace Keys - For this case, config entries need to merged / overwritten depending on the context and meaning of the config key.
Delete Keys - This case must be avoided. Config entry should exist and marked obsolete for at least 1-2 versions and then be deleted.
 

Design specifications:

https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/785448990/CORTX+Software+Upgrade
# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
1: Created build which contains changes of this PR#1840
    cortxcontrol: cortx-docker.colo.seagate.com/seagate/cortx-control:2.0.0-6690-custom-ci
    cortxdata: cortx-docker.colo.seagate.com/seagate/cortx-data:2.0.0-6690-custom-ci
    cortxserver: cortx-docker.colo.seagate.com/seagate/cortx-rgw:2.0.0-6690-custom-ci
    cortxha: cortx-docker.colo.seagate.com/seagate/cortx-control:2.0.0-6690-custom-ci
    cortxclient: cortx-docker.colo.seagate.com/seagate/cortx-data:2.0.0-6690-custom-ci
2: Deploy the singlenode setup with above build. It is up and running.
3: Now we want to upgrade cortx-data:2.0.0-6690-custom-ci to cortx-data:2.0.0-6698-custom-ci
3.1: For this, we entered the docker image of 6690 
docker run -it cortx-docker.colo.seagate.com/seagate/cortx-data:2.0.0-6690-custom-ci /bin/bash

3.2: Edited following lines in file:/opt/seagate/cortx/RELEASE.INFO
VERSION: "2.0.0-6698"
COMPONENTS:
    - "cortx-motr-2.0.0-6698_gita7c53b28.el8.x86_64.rpm"

3.3: Save the docker image with new name
docker commit 704704d8cac0 cortx-docker.colo.seagate.com/seagate/cortx-data:2.0.0-6698-custom-ci

4: Exit the docker image and update the solution.yaml as below
  images:
    cortxdata: cortx-docker.colo.seagate.com/seagate/cortx-data:2.0.0-6698-custom-ci
5: Run upgrade command: 
Clone git clone https://github.com/seagate/cortx-k8s -b cortx-test
./upgrade-cortx-cloud.sh start -p data 

Note: Make sure to edit ./upgrade-cortx-cloud.sh  as below for local docker images
76 function fetch_solution_images() {
 77     printf "Using solution config file '%s'\n" "${SOLUTION_FILE}"
 78     # Fetch Upgrade Images from solution.yaml and validate them
 79     cortxcontrol_image=$(parse_solution 'solution.images.cortxcontrol' | cut -f2 -d'>')
 80     #validate_upgrade_images "${cortxcontrol_image}"
 81     cortxha_image=$(parse_solution 'solution.images.cortxha' | cut -f2 -d'>')
 82     #validate_upgrade_images "${cortxha_image}"
 83     cortxdata_image=$(parse_solution 'solution.images.cortxdata' | cut -f2 -d'>')
 84     cortxserver_image=$(parse_solution 'solution.images.cortxserver' | cut -f2 -d'>')
 85     #validate_upgrade_images "${cortxserver_image}"
 86 }
 

Output After upgrade. If you see pods are going in upgrade sequence

Every 2.0s: kubectl get pods                                                                                                                                     Tue Jun 14 09:16:00 2022

NAME                                                 READY   STATUS        RESTARTS   AGE
cortx-consul-client-jhwxb                            1/1     Running       0          141m
cortx-consul-server-0                                1/1     Running       0          141m
cortx-control-6bbfc4dcf9-7kwjq                       1/1     Running       0          141m
cortx-data-ssc-vm-g2-rhev4-3118-54ccdb45d6-cwmv8     0/4     Init:0/2      0          2s
cortx-data-ssc-vm-g2-rhev4-3118-845656b76b-4925q     4/4     Terminating   0          14m
cortx-ha-68b8b9c5c4-zmvjz                            3/3     Running       0          141m
cortx-kafka-0                                        1/1     Running       0          141m
cortx-server-ssc-vm-g2-rhev4-3118-655c959cd9-cdmr2   2/2     Running       0          141m
cortx-zookeeper-0                                    1/1     Running       0          1

Every 2.0s: kubectl get pods                                                                                                                                     Tue Jun 14 09:16:15 2022

NAME                                                 READY   STATUS        RESTARTS   AGE
cortx-consul-client-jhwxb                            1/1     Running       0          142m
cortx-consul-server-0                                1/1     Running       0          142m
cortx-control-6bbfc4dcf9-7kwjq                       1/1     Running       0          142m
cortx-data-ssc-vm-g2-rhev4-3118-54ccdb45d6-cwmv8     4/4     Running       0          17s
cortx-data-ssc-vm-g2-rhev4-3118-845656b76b-4925q     4/4     Terminating   0          14m
cortx-ha-68b8b9c5c4-zmvjz                            3/3     Running       0          141m
cortx-kafka-0                                        1/1     Running       0          142m
cortx-server-ssc-vm-g2-rhev4-3118-655c959cd9-cdmr2   2/2     Running       0          141m
cortx-zookeeper-0                                    1/1     Running       0          142m

[root@ssc-vm-g2-rhev4-3118 ~]# watch kubectl get pods
Every 2.0s: kubectl get pods                                                                                                                                     Tue Jun 14 09:16:50 2022

NAME                                                 READY   STATUS    RESTARTS   AGE
cortx-consul-client-jhwxb                            1/1     Running   0          142m
cortx-consul-server-0                                1/1     Running   0          142m
cortx-control-6bbfc4dcf9-7kwjq                       1/1     Running   0          142m
cortx-data-ssc-vm-g2-rhev4-3118-54ccdb45d6-cwmv8     4/4     Running   0          52s
cortx-ha-68b8b9c5c4-zmvjz                            3/3     Running   0          141m
cortx-kafka-0                                        1/1     Running   0          142m
cortx-server-ssc-vm-g2-rhev4-3118-655c959cd9-cdmr2   2/2     Running   0          141m
cortx-zookeeper-0                                    1/1     Running   0          142m

Upgrade Successful for cortx-data-ssc-vm-g2-rhev4-3118

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py [hunk 1]
```diff
@@ -383,6 +383,70 @@ def validate_motr_rpm(self):
...
+                found = True
+                break
+        # If not found, append or skip according to flag
+        if not found:
+            if flag == 'append':
+                self.logger.info(f"({k},{v}) not found in config. flag is {flag} so appending.\n")
+                lines.append(f"{k}={v}\n")
+            elif flag == 'update':
+                self.logger.info(f"({k},{v}) not found in config. flag is {flag} so skipping.\n")
+            found = False
```
Comment (yeshpal-jain-seagate) [inline scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py:439]: If flag==delete, than let user know the key is not available in configuration.

File: scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py [hunk 2]
```diff
@@ -383,6 +383,70 @@ def validate_motr_rpm(self):
...
+            lines.append(line)
+    num_lines = len(lines)
+    self.logger.info(f"Before update, num_lines={num_lines}\n")
+
+    #Check for keys in file
+    for (k, v) in kv_list:
+        found = False
+        for lno in range(num_lines):
+            # If found, update inline.
+            if lines[lno].startswith(f"{k}="):
```
[the remark this answers is not in the document]
Comment (atulsdeshmukh2312) [reply scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py:423]: TODO: 
This PR is to verify the upgrade framework oof provisioner.
This trim review comment will be addressed in the following ticket.

File: scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py [hunk 3]
```diff
@@ -383,6 +383,70 @@ def validate_motr_rpm(self):
...
+                lines.append(f"{k}={v}\n")
+            elif flag == 'update':
+                self.logger.info(f"({k},{v}) not found in config. flag is {flag} so skipping.\n")
+            found = False
+
+    num_lines = len(lines)
+    self.logger.info(f"After update, num_lines={num_lines}\n")
+
+    # Write buffer to file
+    with open(f"{MOTR_M0D_CONF_FILE}", "w+") as fp:
```
[the remark this answers is not in the document]
Comment (atulsdeshmukh2312) [reply scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py:445]: This is related to consistency of config file while writing. Will be covered in the following ticket.

**Review by yeshpal-jain-seagate — APPROVED** [review]
Because this is a PR to test provisioner framework, all the required comments need to be addressed in follow up implementation PR

**Review by yatin-mahajan — APPROVED** [review]
Looks good to me.

File: scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py [hunk 4]
```diff
@@ -383,6 +383,61 @@ def validate_motr_rpm(self):
...
+            found = False
+        else:
+            if flag == 'append':
+                self.logger.error(f"({k},{v}) found in config. so skipping {flag}.\n")
+    num_lines = len(lines)
+    self.logger.info(f"After update, num_lines={num_lines}\n")
+
+    # Write buffer to file
+    # TODO: Consistency whould be maintained while writing to file.
+    # For example, if upgrade if crashed while updating the file then use the previous consistent copy.
```
Comment (madhavemuri) [inline scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py:435]: s/upgrade if/upgrade is

File: scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py [hunk 5]
```diff
@@ -1512,3 +1565,91 @@ def start_service(self, service, idx):
     cmd = f"{MOTR_SERVER_SCRIPT_PATH} m0d-{fid}"
     execute_command_console(self, cmd)
     return
+
+# Itertate recursiverly through dictionary and extract leaf <key, value>
```
Comment (madhavemuri) [inline scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py:1569]: s/Itertate recursiverly/Iterate recursively/

**Pull request merged**